compiler/src/dmd/expressionsem.d

```diff
@@ -9958,8 +9951,8 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
                 else
                 {
                     Expression e0;
-                    // If ae.e2 is not a variable, construct a temp variable, as _d_arraysetctor requires `ref` access
-                    if (!ae.e2.isVarExp)
+                    // promote an rvalue RHS element to a temporary, it's passed by ref to _d_arraysetctor
+                    if (!ae.e2.isLvalue)
```

Comment:
[This change is tested in line 66 (`S[3] fromSingleLval = *p;`), getting rid of an extra copy if the RHS is a non-var lvalue.]